### PR TITLE
uboot: add support for the ODROID-C2

### PR DESCRIPTION
**Pull request by lopsided98**
###### Motivation for this change

This adds support for the ODROID-C2 to U-Boot in nixpkgs. It minimizes the use of binary blobs and vendor tools. It has been tested in combination with mainline Linux 5.4.1.

The Amlogic S905 uses 5 bootloader stages. The first (`bl1.bin.hardkernel`) is partially placed in the MBR bootstrap code area and partially in the sector after the MBR. The rest of the stages are stored in a FIP image placed at sector 97. `bl30.bin` and `bl301.bin` are binary blobs ([sources are technically available for `bl301.bin`](https://github.com/hardkernel/u-boot_firmware/tree/odroidc2-bl301), but building them is more trouble than it's worth). The other two images are ATF and U-Boot, both of which have upstream support for the SoC/board.

Hardkernel's U-Boot tree has a custom version of the `fip_create` tool taken from ATF. It was modified to add padding to the BL301 image (to make room for a signature) and place the BL301 image between BL30 and BL31 (ATF) with a custom UUID. BL301 needs to be signed by the `aml_encrypt_gxb` tool which is only provided as a binary for `x86_64-linux`.

Upstream ATF has replaced `fip_create` with `fiptool`. It is possible to add images with custom UUIDs using `fiptool`, but they are always added to the end of the file. This is a problem because the bootloader blob responsible for loading the FIP only looks at the order of the images, not their UUIDs. `fiptool` [can be patched to place the BL301 image at the right position](https://github.com/afaerber/arm-trusted-firmware/commit/3da46db507b23e6dfb1ffaa033fa3f7392f1437e), but it turns out that this is not really necessary. Since the bootloader doesn't care about UUIDs, we can simply put the images in the right order with the wrong UUIDs using an unmodified `fiptool`. See https://github.com/afaerber/meson-tools/issues/3 for more information.

The [meson-tools](https://github.com/afaerber/meson-tools) project contains a reverse engineered version of `aml_encrypt_gxb`, which can be used instead.

There is an opportunity for bikeshedding here: what should the ATF package be called?
* armTrustedFirmwareS905
* armTrustedFirmwareAmlogicS905
* armTrustedFirmwareMesonGxbb

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @samueldr @dezgeg @grahamc

### CODE REVIEW

File: pkgs/misc/uboot/default.nix [hunk 1]
```diff
@@ -163,6 +163,51 @@ in {
...
+  # dd if=bl1.bin.hardkernel of=<device> conv=fsync bs=1 count=442
+  # dd if=bl1.bin.hardkernel of=<device> conv=fsync bs=512 skip=1 seek=1
+  # dd if=u-boot.gxbb of=<device> conv=fsync bs=512 seek=97
+  ubootOdroidC2 = let
+    firmwareBlobs = fetchFromGitHub {
+      owner = "armbian";
+      repo = "odroidc2-blobs";
+      rev = "47c5aac4bcac6f067cebe76e41fb9924d45b429c";
+      sha256 = "1ns0a130yxnxysia8c3q2fgyjp9k0nkr689dxk88qh2vnibgchnp";
+    };
```
Comment (samueldr) [inline pkgs/misc/uboot/default.nix]: At the very 
```diff
       sha256 = "1ns0a130yxnxysia8c3q2fgyjp9k0nkr689dxk88qh2vnibgchnp";
+      meta.license = lib.licenses.unfreeRedistributableFirmware;
     };
```

**Review by samueldr — COMMENTED** [review]
I don't have the HW to verify. Otherwise looks good to me, except for the small change.

👍 It cross-compiles!

I think it's also appropriate to add yourself to the u-boot maintainers :).

**Comment by lopsided98** [comment]
I added the license and fixed the conflicts with my Rock64 changes.

**Review by samueldr — APPROVED** [review]

**Review by markuskowa — APPROVED** [review]
This is great!

**Pull request merged**